bend_gcode.py

```diff
@@ -5,135 +5,91 @@
-    thisLine = re.compile('(?i)^[gG][0-3](?:\s+x(?P<x>-?[0-9.]{1,15})|\s+y(?P<y>-?[0-9.]{1,15})|\s+z(?P<z>-?[0-9.]{1,15})|\s+e(?P<e>-?[0-9.]{1,15})|\s+f(?P<f>-?[0-9.]{1,15}))*')
-    lineEntries = thisLine.match(currentLine)
-    if lineEntries:
-        return GCodeLine(lineEntries.group('x'), lineEntries.group('y'), lineEntries.group('z'), lineEntries.group('e'), lineEntries.group('f'))
-
-def writeLine(G, X, Y, Z, F = None, E = None): #write a line to the output file
-    outputSting = "G" + str(int(G)) + " X" + str(round(X,5)) + " Y" + str(round(Y,5)) + " Z" + str(round(Z,3))
-    if E is not None:
-        outputSting = outputSting + " E" + str(round(float(E),5))
-    if F is not None:
-        outputSting = outputSting + " F" + str(int(float(F)))
-    outputFile.write(outputSting + "\n")
-
-"""
-# legacy - toooo slow!    
-def onSplineLength(Zheight) -> float: #calculates a new z height if the spline is followed
-    return Zheight #for debugging
-    discretizationLength = 0.01 #Steps taken to find the spline height
-    currentHeight = 0.00
-    currentLength = 0.00
-    while currentLength < Zheight:
-        currentLength += np.sqrt((SPLINE(currentHeight)-SPLINE(currentHeight+discretizationLength))**2 + discretizationLength**2)
-        currentHeight += discretizationLength
-    return currentHeight
-"""
-
-def onSplineLength(Zheight) -> float: #calculates a new z height if the spline is followed
-    for i in range(len(SplineLookupTable)):
-        height = SplineLookupTable[i]
-        if height >= Zheight:
-            return i * DISCRETIZATION_LENGTH
-    print("Error! Spline not defined high enough!")
-
-def createSplineLookupTable():
-    heightSteps = np.arange(DISCRETIZATION_LENGTH, SPLINE_Z[-1], DISCRETIZATION_LENGTH)
-    for i in range(len(heightSteps)):
-        height = heightSteps[i]
-        SplineLookupTable.append(SplineLookupTable[i] + np.sqrt((SPLINE(height)-SPLINE(height-DISCRETIZATION_LENGTH))**2 + DISCRETIZATION_LENGTH**2))
-    
-
-
-lastPosition = Point2D(0, 0)
-currentZ = 0.0
-lastZ = 0.0
-currentLayer = 0
-relativeMode = False
-createSplineLookupTable()
-
-with open(INPUT_FILE_NAME, "r") as gcodeFile, open(OUTPUT_FILE_NAME, "w+") as outputFile:
-        for currentLine in gcodeFile:
+@dataclass
+class Point2D:
+    x: float
+    y: float
+
+@dataclass
+class GCodeLine:
+    x: float
+    y: float
+    z: float
+    e: float
+    f: int
+
+    gcode_format: ClassVar[re.Pattern] = (
+        re.compile('(?i)^[gG][0-3](?:\s'
+                   '+x(?P<x>-?[0-9.]{1,15})|\s'
+                   '+y(?P<y>-?[0-9.]{1,15})|\s'
+                   '+z(?P<z>-?[0-9.]{1,15})|\s'
+                   '+e(?P<e>-?[0-9.]{1,15})|\s'
+                   '+f(?P<f>-?[0-9.]{1,15}))*')
+    )
+
+    @classmethod
+    def from_gcode(cls, line: str):
+        line_entries = cls.gcode_format.match(line)
+        if line_entries:
+            return cls(*line_entries.groups())
```

Comment:
This is very slick. Great work!

Reply:
Stefan made the match string, I just made it a bit more readable, and cleaned up how it's used :-)